### PR TITLE
Implement np.intersect1d

### PR DESCRIPTION
**Pull request by aldragan0**
This is an implementation of [np.intersect1d](https://numpy.org/devdocs/reference/generated/numpy.intersect1d.html) from [#70 ](https://github.com/google/jax/issues/70) (and [#2078 ](https://github.com/google/jax/issues/2078))

### CODE REVIEW

File: jax/numpy/lax_numpy.py [hunk 1]
```diff
@@ -1243,6 +1243,57 @@ def in1d(ar1, ar2, assume_unique=False, invert=False):
...
+@partial(jit, static_argnums=2)
+def _intersect1d_sorted_mask(ar1, ar2, return_indices=False):
+    """
+    Helper function for intersect1d which is jit-able
+    """
+    ar = concatenate((ar1, ar2))
+
+    if return_indices:
+      indices = argsort(ar)
+      aux = ar[indices]
```
Comment (jakevdp) [inline jax/numpy/lax_numpy.py:1255]: It's possible to use `lax.sort` directly to get the sorted array and the indices in a single pass (if I recall correctly, argsort uses this and then discards the sorted array).

**Review by jekbradbury — APPROVED** [review]

**Review by jakevdp — COMMENTED** [review]
Looks great, thanks! One comment about a potential performance improvement.

**Comment by jekbradbury** [comment]
I think Jake and I were reviewing at the same time! His suggestion sounds good; maybe you can make another PR to incorporate it?

**Comment by aldragan0** [comment]
Sure, I'll add it and link to this PR

**Comment by jakevdp** [comment]
Correct, the semantics of numpy's `intersect1d` are incompatible with JAX transformations like `jit`.

#3614 tracks the effort to add size hints to functions that share this issue – many are done but `intersect1d` is not (it's been a while, but I recall one reason I didn't tackle it is that it wasn't clear to me what a natural fill value should be in the case that `size` is larger than the true intersection).

Many of these kinds of features are implemented on an as-needed basis. If this is important to your use-case, you might chime-in on that issue with ideas of what kind of shape-hint behavior would be useful to you.

**Pull request merged**